enhancements/network/allow-mtu-changes.md

```diff
@@ -0,0 +1,285 @@
+negotiate their segment size based on the lower MTU of the endpoints on 
+connection.
+
+So generally, different MTUs on endpoints affect ongoing connection-oriented
+traffic or connection-less traffic, when the known destination MTU is not the
+actual destination MTU. In this case, the most likely scenario is that traffic
+is dropped on the receiving end by OVS if larger than the destination MTU.
+
+There are circumstances that prevent an endpoint from being aware of the actual
+MTU to a destination, which depends on Path MTU discovery and specific ICMP
+`FRAG_NEEDED` messages:
+* A firewall is blocking these ICMP messages or the ICMP messages are not being
+  relayed to the endpoint.
+* There is no router between the endpoints generating these ICMP messages.
+
+Let's look at different scenarios.
+
+##### Node to node
+
+On the receiving end, a nic driver might size the buffers in relation to the
+configured MTU and drop larger packets before they are handed off to the system.
+
+Past that, OVN-K sets up flows in OVS br-ex for packets that are larger than pod
+MTU and sends them off to the network stack to generate ICMP `FRAG_NEEDED`
+messages. If these packets exceed  the MTU of br-ex, they will be dropped by OVS
+and never reach the network stack. Otherwise they will reach the network stack
+but not generate ICMP `FRAG_NEEDED` messages as network stack only does so for
+traffic being forwarded and not for traffic with that node as final destination.
+
+As there is generally no router in between two cluster nodes, more than likely a
+node would not be aware of the path MTU to another node.
+
+##### Node to pod
+
+As explained before, network stack receives larger packets betwewen host MTU and
+pod MTU and might cause ICMP `FRAG_NEEDED` messages to be sent to the originating
+node such that a node might be aware of the proper path MTU when reaching out to
+pod. Otherwise larger than pod MTU traffic will dropped by OVS.
+
+##### Pod to Node
+
+On this datapath, OVS at the destination node will drop the larger packets without
+generating ICMP `FRAG_NEEDED` messages as the node is the final destination of the
+traffic. The originating pod is never aware of the actual path MTU.
+
+##### Pod to Pod
+
+This traffic is encapsulated with geneve. The geneve driver might drop it and
+generate ICMP `FRAG_NEEDED` messages back to the originating pod if it is trying
+to send packets that would not fit in the originating node MTU once encapsulated.
+But OVN is not prepared to relay back these ICMP messages to the originating pod
+so it would not be aware of an appropiate MTU to use.
+
+On the receiving end, OVS would drop the packet silently if larger than the
+destination MTU of the veth interface. Even if this would not be the case, the
+veth driver itself would drop the packet silently if over the MTU of the pod's
+end veth interface.
+
+## Alternatives
+
+### New ovn-k setting: `routable-mtu`
```

Comment:
So this sounds better than the proposed solution... why aren't we doing it this way?

Reply:
Yes, I was concerned that 2 reboots would not be acceptable from a customer perspective. @mcurry-rh What are your thoughts on having to perform 2 reboots to change the mtu?

Reply:
Replicating the feedback we got form @mcurry-rh 
`not ideal...acceptable...MTU adjustment is a rare event, so 2 reboots, while painful, is not fatal and achieves the objective`
So the **second alternative** is based on already available node maintenance knowledge, simpler to implement and a safer approach all around while the **main alternative** is more efficient at the cost of that safety. We could prototype as well.
@abhat @trozet @knobunc @dcbw we would need to make a call on this. Do you have any opinion?

Reply:
So, if the cluster is actually "broken" because of the bad MTU, then having to do two reboots isn't that bad since you're probably not running anything useful anyway.

And if it's not broken, then the MTU change probably isn't _urgent_, and the procedure doesn't actually require that the two reboots happen back-to-back; they could happen 24 hours apart or something. (Right? The cluster is stable/consistent in the inter-reboot phase?) So we could even just make it so that the CNO doesn't initiate any rolling reboots itself, it just does:

- CNO makes the initial change to mtu/routable-mtu
- CNO observes nodes until it sees that every node has rebooted (for whatever reason) and is using the changed configuration.
- CNO makes the second change to mtu/routable-mtu
- CNO observes nodes until it sees that every node has rebooted and is using the changed configuration.
- CNO updates the operator status accordingly

So then the admin could schedule two sets of rolling reboots on consecutive nights, or even just make the config change and then forget about it, and the first change would complete the next time they did a z-stream update and the second change would complete after the next update after that.

Reply:
>  (Right? The cluster is stable/consistent in the inter-reboot phase?) 
Yes.

Reply:
"So, if the cluster is actually "broken" because of the bad MTU" -- That is not always a safe assumption.  One case we had was where a customer had a large, running cluster and wanted to add new nodes.  But the new nodes were on OpenShift and they needed to drop the MTU to allow for the VxLAN header in the OSP networking.  I assume most cases will be like that, otherwise they could just reinstall...

Reply:
>"So, if the cluster is actually "broken" because of the bad MTU" -- That is not always a safe assumption.

Hence the "if"